Add Data Fetching Guide

amirhhashemi opened the pull request:
- [x] I have read the [Contribution guide](https://github.com/solidjs/solid-docs/blob/main/CONTRIBUTING.md)
- [x] This PR references an issue (except for typos, broken links, or other minor problems)

### Description

The idea of the Guides section is to provide a resource for developers who don't read Solid.js and Solid Router docs before learning SolidStart, such as React developers who may think of SolidStart as what Next.js is for React and start learning Solid.js from SolidStart. However, I think a Guides section can be useful to a lot of people regardless of whether they come from React.

The goal is to get those users (or anyone really) up and running with SolidStart as quickly as possible, while also guiding them to read more relevant docs if they want to learn more.

This PR adds a Data Fetching page under the Guides section that covers basic data fetching patterns while providing links to relevant documentation for those who want to learn more.

See https://github.com/solidjs/solid-docs/issues/861 for more context. 

### Related issues & labels

#914 https://github.com/solidjs/solid-docs/issues/861

amirhhashemi committed:
First draft

LadyBluenotes commented:
Hi @amirhhashemi,

First of all, thank you for taking the time to work on this, I appreciate the time you're taking out of your day to improve our documentation.  

It might be useful to take a look at the [diataxis documentation on how guides should be written](https://diataxis.fr/how-to-guides/). We use this as a reference for writing our docs, and i feel like it could be helpful to consider for this!

If you'd like some more directed feedback, however, please let me know

amirhhashemi committed:
iteration #1

amirhhashemi commented:
Hi @LadyBluenotes 

Thank you for the feedback. I made some changes based on the diataxis guide:
1. I removed all explanations that are more suited for a tutorial page and only relied on links to provide more info.
2. I tried to make each section more action-oriented. Now, the titles start with "How to" and the minimal description of each section starts with "To do".
3. I reordered the sections to create a better flow.

I'm not sure about:
1. Whether I should keep the examples self-contained. I think it's a good thing because readers can pick up each section and copy its example without needing to read the previous sections.
2. Whether the section about Tanstack Query belongs on this page. Maybe it's more suited for a tutorial page or the "build your application" section.

LadyBluenotes commented:
> Hi @LadyBluenotes
> 
> Thank you for the feedback. I made some changes based on the diataxis guide:
> 
>     1. I removed all explanations that are more suited for a tutorial page and only relied on links to provide more info.
> 
>     2. I tried to make each section more action-oriented. Now, the titles start with "How to" and the minimal description of each section starts with "To do".
> 
>     3. I reordered the sections to create a better flow.
> 
> 
> I'm not sure about:
> 
>     1. Whether I should keep the examples self-contained. I think it's a good thing because readers can pick up each section and copy its example without needing to read the previous sections.
> 
>     2. Whether the section about Tanstack Query belongs on this page. Maybe it's more suited for a tutorial page or the "build your application" section.

Thank you for adding some changes! One comment I do want to make, is that I think adding back some (basic) explanations can be helpful. Doesn't have to explain anything but the process, such as this:

> To show users that data is loading, you can display a loading UI while they wait. You can achieve this by wrapping your component in [Solid's `<Suspense>` component](https://docs.solidjs.com/reference/components/suspense) and using the `fallback` prop to specify what should be shown during the loading process. Here's how:
> 
>  1.  Import the <Suspense> component from SolidJS.
>  2.  Wrap your component with <Suspense> and define a fallback UI to display while waiting for the data to load.

This way it follows we have the code snippet explained a bit in case users need some kind of explanation (but not giving too much background about the topic itself, which is more appropriate for other sections in the docs). This is a pretty rough example but we can either have it as a quick list (if you feel it's necessary) or just a simple explanation of  how things are set up in the snippet.

I do agree that, to an extent, having each section self-contained is important. We don't want them so far removed that they could be considered separate entirely, but enough that there is some relationship (even if marginally).

As for the Tanstack Query section, I do agree that it would probably be better suited for a tutorial than in here. That being said, a callout could probably replace the section saying something as simple as what you wrote for the first sentence:

> [Tanstack Query](https://tanstack.com/query/latest) can be used for more advanced features such as automatic background re-fetching or infinite queries.

Do you have any questions/ comments about what I mentioned?

amirhhashemi committed:
iteration 2

amirhhashemi commented:
Thank you for the feedback.

I've added basic explanations for each section. However, I have some concerns:

1. Some explanations may be redundant and could be streamlined
2. I'm particularly uncertain about the `createResource` section - it's challenging to explain without diving too deep into its internal workings

I would appreciate your feedback on these points. Please feel free to edit the document directly if needed.

Regarding Tanstack Query: I've removed the dedicated section and added a brief note about it at the end of the `createResource` section instead.

amirhhashemi committed:
fix typo

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

amirhhashemi commented:
I'm wondering whether we should remove the "how to" from section titles:

- How to handle errors -> Handling errors
- How to preload data -> Preloading data
etc

The way this guide is written, it's obvious we are only talking about *how* to do stuff, and not why and when. I think we can remove it. What do you think?

LadyBluenotes commented:
@amirhhashemi Thank you for adding the explanations! I'll try and get to reviewing them this weekend (and seeing how we can approach the createResource section) but I wanted to say I think it's a great Idea to update the titles to remove the `how to` parts!

amirhhashemi committed:
iteration 3

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

LadyBluenotes commented on src/routes/solid-start/guides/data-fetching.mdx:
````diff
@@ -0,0 +1,227 @@
+}
+```
+
+In this example, we created a [`query`](https://docs.solidjs.com/solid-router/reference/data-apis/query) and used the [`createAsync`](https://docs.solidjs.com/solid-router/reference/data-apis/create-async) primitive to access the query data inside our component.
````
```suggestion
In this example, a [`query`](/solid-router/reference/data-apis/query) is created. In order to access it's data within the component, the [`createAsync`](/solid-router/reference/data-apis/create-async) primitive was used.
```

LadyBluenotes commented on src/routes/solid-start/guides/data-fetching.mdx:
```diff
@@ -0,0 +1,227 @@
+- Solid Router [`preload`](https://docs.solidjs.com/solid-router/reference/preload-functions/preload)
+- Solid [`createResource`](https://docs.solidjs.com/reference/basic-reactivity/create-resource)
+- Solid [`<Suspense>`](https://docs.solidjs.com/reference/components/suspense)
+- Solid [`<ErrorBoundary>`](https://docs.solidjs.com/reference/components/error-boundary)
```
I'm not entirely sure this section is necessary since each part is linked out when used

LadyBluenotes commented on src/routes/solid-start/guides/data-fetching.mdx:
```diff
@@ -0,0 +1,227 @@
+
+## Handling errors
+
+To show a fallback UI in case data fetching fails, you can use [Solid `<ErrorBoundary>`](https://docs.solidjs.com/reference/components/error-boundary):
```
```suggestion
If the data fetching fails, a fallback UI can be displayed using an [`<ErrorBoundary>`](/reference/components/error-boundary):
```

LadyBluenotes commented on src/routes/solid-start/guides/data-fetching.mdx:
```diff
@@ -0,0 +1,227 @@
+
+## Showing loading UI
+
+To show a loading UI while data is loading, you can use [Solid `<Suspense>`](https://docs.solidjs.com/reference/components/suspense):
```
```suggestion
To display a loading UI while data is being fetched, you can use [Solid `<Suspense>`](/reference/components/suspense):
```

LadyBluenotes commented on src/routes/solid-start/guides/data-fetching.mdx:
```diff
@@ -0,0 +1,227 @@
+
+## Preloading data
+
+To optimize data fetching during user navigation, you can use [`preload`](https://docs.solidjs.com/solid-router/reference/preload-functions/preload) to preload a query:
```
```suggestion
Data fetching can be optimized during user navigation through preloading the data with the [`preload`](/solid-router/reference/preload-functions/preload) function: 
```

LadyBluenotes commented on src/routes/solid-start/guides/data-fetching.mdx:
```diff
@@ -0,0 +1,227 @@
+
+## Passing parameters to queries
+
+To create a query that accepts parameters, you can pass any number of arguments to your query function:
```
```suggestion
When creating a query that accepts parameters, you can define your query function to take any number of arguments:
```

LadyBluenotes commented on src/routes/solid-start/guides/data-fetching.mdx:
````diff
@@ -0,0 +1,227 @@
+}
+```
+
+In this example, we used [`useParams`](https://docs.solidjs.com/solid-router/reference/primitives/use-params) to get the `id` path parameter and passed it to the `getPost` query to get the data for that specific post.
````
```suggestion
In this example, [`useParams`](/solid-router/reference/primitives/use-params) was used to access the `id` path parameter. You can then pass it to the `getPost` query to get the data for that specific post.
```

LadyBluenotes commented on src/routes/solid-start/guides/data-fetching.mdx:
```diff
@@ -0,0 +1,227 @@
+
+## Using database or an ORM
+
+To directly use your database or ORM in a query, you should make the query server-only so your database credentials are not exposed to the client. You can do that by adding [`"use server"`](https://docs.solidjs.com/solid-start/reference/server/use-server) to the first line of your query function:
```
```suggestion
To safely interact with your database or ORM in a query, ensure the query is server-only to prevent exposing your database credentials to the client. You can do this by adding [`"use server"`](/solid-start/reference/server/use-server) as the first line of your query function:
```

LadyBluenotes commented on src/routes/solid-start/guides/data-fetching.mdx:
```diff
@@ -0,0 +1,227 @@
+
+## Fetching data on the client
+
+To fetch data only on the client side, you can use the [`createResource`](https://docs.solidjs.com/reference/basic-reactivity/create-resource) primitive:
```
```suggestion
When you only need to fetch data on the client side, you can use the [`createResource`](/reference/basic-reactivity/create-resource) primitive:
```

LadyBluenotes commented on src/routes/solid-start/guides/data-fetching.mdx:
````diff
@@ -0,0 +1,227 @@
+}
+```
+
+Read more in the [`createResource`](https://docs.solidjs.com/reference/basic-reactivity/create-resource) docs.
````
```suggestion
You can [read more about `createResource` here](/reference/basic-reactivity/create-resource) .
```

LadyBluenotes commented on src/routes/solid-start/guides/data-fetching.mdx:
```diff
@@ -0,0 +1,227 @@
+Read more in the [`createResource`](https://docs.solidjs.com/reference/basic-reactivity/create-resource) docs.
+
+<Callout type="info">
+    [Tanstack Query](https://tanstack.com/query/latest/docs/framework/solid/overview) can be used for more advanced features such as automatic background re-fetching or infinite queries.
```
```suggestion
    For advanced features like automatic background re-fetching or infinite queries, you can use [Tanstack Query.](https://tanstack.com/query/latest/docs/framework/solid/overview)
```

LadyBluenotes commented on src/routes/solid-start/guides/data-fetching.mdx:
```diff
@@ -0,0 +1,227 @@
+
+Read more in the [`createResource`](https://docs.solidjs.com/reference/basic-reactivity/create-resource) docs.
+
+<Callout type="info">
```
```suggestion
<Callout type="info" title="Advanced Data Handling">
```

LadyBluenotes reviewed (changes requested):
Let me know how you feel about these changes, but I feel like you did a great job explaining each code snippet briefly.

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

amirhhashemi committed:
Update src/routes/solid-start/guides/data-fetching.mdx

Co-authored-by: Sarah <gerrardsarah@gmail.com>

amirhhashemi committed:
Update src/routes/solid-start/guides/data-fetching.mdx

Co-authored-by: Sarah <gerrardsarah@gmail.com>

amirhhashemi committed:
Update src/routes/solid-start/guides/data-fetching.mdx

Co-authored-by: Sarah <gerrardsarah@gmail.com>

amirhhashemi committed:
Update src/routes/solid-start/guides/data-fetching.mdx

Co-authored-by: Sarah <gerrardsarah@gmail.com>

amirhhashemi committed:
Update src/routes/solid-start/guides/data-fetching.mdx

Co-authored-by: Sarah <gerrardsarah@gmail.com>

amirhhashemi committed:
Update src/routes/solid-start/guides/data-fetching.mdx

Co-authored-by: Sarah <gerrardsarah@gmail.com>

amirhhashemi committed:
Update src/routes/solid-start/guides/data-fetching.mdx

Co-authored-by: Sarah <gerrardsarah@gmail.com>

amirhhashemi committed:
Update src/routes/solid-start/guides/data-fetching.mdx

Co-authored-by: Sarah <gerrardsarah@gmail.com>

amirhhashemi committed:
Update src/routes/solid-start/guides/data-fetching.mdx

Co-authored-by: Sarah <gerrardsarah@gmail.com>

amirhhashemi committed:
Update src/routes/solid-start/guides/data-fetching.mdx

Co-authored-by: Sarah <gerrardsarah@gmail.com>

amirhhashemi committed:
Update src/routes/solid-start/guides/data-fetching.mdx

Co-authored-by: Sarah <gerrardsarah@gmail.com>

amirhhashemi committed:
iteration 4

amirhhashemi commented:
I really appreciate your feedback and help @LadyBluenotes. I think this PR if ready for the final review.

~I can work on a data mutation guide in another PR if you like.~ I just went for it:) https://github.com/solidjs/solid-docs/pull/994

amirhhashemi commented:
After this is merged, I see no reason for the current Date Loading page to remain. Should I remove it @LadyBluenotes?

Edit: I created a PR that removes this page https://github.com/solidjs/solid-docs/pull/995

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

amirhhashemi committed:
iteration 5

amirhhashemi committed:
fix typo

devagrawal09 commented on src/routes/solid-start/guides/data-fetching.mdx:
````diff
@@ -0,0 +1,218 @@
+}
+```
+
+You can [read more about `createResource` here](/reference/basic-reactivity/create-resource).
````
~So `createAsync` can still be used to fetch on the client side, and the only reason you'd use `createResource` instead is if you're not using `solid-router`, not if you're fetching on the client. This section seems to confuse those two things together.~

~I'd break this into two separate sections, one that shows fetching on the client with `createAsync` without `"use server"`, and another that shows fetching using `createResource` on both client and server.~

Nevermind, I just noticed that most of your `createAsync` snippets are fetching on the client already. The only correction I have is this section should be called `"Fetching data without Solid Router"` or something like that.

amirhhashemi replied:
From the [diataxis](https://diataxis.fr/how-to-guides/#how-to-guides-addressed-to-problems) guide that we use as reference:
> How-to guides must be written from the perspective of the user, not of the machinery. A how-to guide represents something that someone needs to get done. It’s defined in other words by the needs of a user. Every how-to guide should answer to a human project, in other words. It should show what the human needs to do, with the tools at hand, to obtain the result they need.
>
> This is in strong contrast to common pattern for how-to guides that often prevails, in which how-to guides are defined by operations that can be performed with a tool or system. The problem with this latter pattern is that it offers little value to the user; it is not addressed to any need the user has. Instead, it’s focused on the tool, on taking the machinery through its motions.

Keeping each section action-oriented, focusing on a practical action that the user might want to do offers better value to the user. If the user wants to learn about a specific tool, it can go to the API Reference.

I believe the documentation should include a section discussing the pros and cons of using Solid.js versus Solid Router APIs for data fetching. Additionally, it would be helpful to clarify why some APIs are part of Solid Router while others belong to Solid.js. However, this page is not aimed at that type of content. Perhaps a tutorial on data fetching would be a more appropriate format for this discussion.

amirhhashemi replied:
In my initial draft, I included a call-out that gave a little hint:

```
<Callout type="info">
  Note that `query` and `createAsync` are imported from `@solidjs/router`. By default, SolidStart uses [Solid Router](https://docs.solidjs.com/solid-router) under the hood. That means we can take advantage of all Solid Router data fetching primitives in SolidStart.
</Callout>
```

This may slightly disrupt the page's theme, but I think reintroducing a call-out like this could significantly help users navigate the material with greater clarity. At least until we add a comprehensive tutorial.

atilafassina replied:
I came here to say exactly that.
While the guide is straight to the point and practical, I think it leaves loose ends with the `query` and the `createAsync` x `createResource` nuances.

A callout would be nice, but I'd also suggest edits on your proposal above. Solid-Router isn't "under the hood" it's quite on the surface and we actually have work in supporting other client-side routers, so I'd suggest removing that piece. Here's my proposal.

I'd recommend putting such a callout in the very first paragraph.

> Solid-Router has primitives to improve deduplication of resources and requests. Wrapping your fetcher in a `query()` will allow for that. And triggering your fetcher with `createAsync` instead of `createResource` will make sure the reactive system tracks it properly. Except for the deduplication mechanism, `createAsync` and `createResource` are interchangeable.

amirhhashemi replied:
@atilafassina The problem with your proposal is that it provides no context about the relationship between SolidStart, Solid, and Solid Router. It only scratches the surface of the primitives and capabilities that Solid and Solid Router provide, which I think is outside the scope of this page.

However, I've updated the introduction. What do you think?

LadyBluenotes replied:
Going to mark this as resolved for the time being so I can merge this guide. It's probably best at this point to bring it up in a separate issue - @amirhhashemi do you mind making it and summarizing the conversation so we can continue the situation there?

devagrawal09 commented on src/routes/solid-start/guides/data-fetching.mdx:
```diff
@@ -0,0 +1,218 @@
+
+<Callout type="info" title="Advanced Data Handling">
+	For advanced features like automatic background re-fetching or infinite queries, you can use [Tanstack Query.](https://tanstack.com/query/latest/docs/framework/solid/overview)
+</Callout>
```
I like the callout, however I still think it would be helpful to show an example snippet here that shows using tanstack query with `"use server"`. Would love to get some more thoughts on this.

LadyBluenotes replied:
I almost wonder if we need a tutorial geared towards advanced data fetching. Like something that goes into the weeds of the more complicated parts people may run into?

amirhhashemi replied:
I think showcasing the usage of Tanstack Query with `"use server"` is interesting enough to include in this guide. But, my concern is that it might go too far. If we want to cover Tanstack Query properly, we should add:

1. A basic example of `createQuery` (with and without `"use server"`)
2. A section about preloading data
3. A section about error handling and loading UI
4. A few examples of advanced things that Tanstack Query can do and SolidStart can't

It's too much for this page. I think it's better to leave it to the user to go through the Tanstack Query docs and figure things out. If there's any ambiguity regarding the use of `"use server"` with Tanstack Query, we should consider updating other sections of the documentation for clarity.

LadyBluenotes replied:
I'd maybe say this is more something for the Tanstack docs than our own. Let me bring this up to some people within the team, as well as Tanner, and see if we can better hammer this out.

LadyBluenotes replied:
There's some stuff incoming for Tanstack - going to resolve this conversation for the time being but we will have to revisit this in the future.

devagrawal09 commented on src/routes/solid-start/guides/data-fetching.mdx:
````diff
@@ -0,0 +1,218 @@
+		</div>
+	);
+}
+```
````
Should we explain why preloading is even needed?

LadyBluenotes replied:
It might be more worthwhile having that in a the other aspects of the docs (like the core parts of "learn" or a "tutorial". This is geared more towards people asking "how to" instead of "why" and i worry if we add the "why" here, it'd be assumed the rest of it should have it as well.

devagrawal09 replied:
that makes sense, we probably need a separate document for explaining the concepts and issues around data fetching + mutation

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

amirhhashemi committed:
format

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

amirhhashemi committed:
update introductin

amirhhashemi committed:
end lists with dot

amirhhashemi committed:
iteration

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

atilafassina commented on src/routes/solid-start/guides/data-fetching.mdx:
````diff
@@ -0,0 +1,211 @@
+
+When creating a query that accepts parameters, define your query function to take any number of arguments:
+
+```tsx {4} {10} {15} title="src/routes/posts/[id]/index.tsx"
+import { ErrorBoundary } from "solid-js";
+import { query, createAsync, type RouteDefinition } from "@solidjs/router";
+
+const getPost = query(async (id: string) => {
+	const post = await fetch(`https://my-api.com/posts/${id}`);
+	return await post.json();
+}, "post");
+
+export const route = {
+	preload: ({ params }) => getPost(params.id),
+} satisfies RouteDefinition;
+
+export default function Page() {
+	const postId = 1;
+	const post = createAsync(() => getPost(postId));
+	return (
+		<div>
+			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
+				<h1>{post().title}</h1>
+			</ErrorBoundary>
+		</div>
+	);
+}
+```
````
Easy to copy/paste snippets from docs is important, and though we're a TypeScript-first resource, it's important not to forget JavaScript users. 

Whenever there are type definitions in a snippet, it's good to provide a tab with a JS-only alternative.

(I had to escape some characters, so I'm not sure this suggestion will really work)

```suggestion
<TabsCodeBlocks>
  <div id="TypeScript">
    ```tsx {4} {10} {15} title="src/routes/posts/[id]/index.tsx"
import { ErrorBoundary } from "solid-js";
import { query, createAsync, type RouteDefinition } from "@solidjs/router";

const getPost = query(async (id: string) => {
	const post = await fetch(`https://my-api.com/posts/${id}`);
	return await post.json();
}, "post");

export const route = {
	preload: ({ params }) => getPost(params.id),
} satisfies RouteDefinition;

export default function Page() {
	const postId = 1;
	const post = createAsync(() => getPost(postId));
	return (
		<div>
			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
				<h1>{post().title}</h1>
			</ErrorBoundary>
		</div>
	);
}
\`\`\`
  </div>
  <div id="JavaScript">
    ```tsx {4} {10} {15} title="src/routes/posts/[id]/index.tsx"
import { ErrorBoundary } from "solid-js";
import { query, createAsync, type RouteDefinition } from "@solidjs/router";

const getPost = query(async (id: string) => {
	const post = await fetch(`https://my-api.com/posts/${id}`);
	return await post.json();
}, "post");

export const route = {
	preload: ({ params }) => getPost(params.id),
};

export default function Page() {
	const postId = 1;
	const post = createAsync(() => getPost(postId));
	return (
		<div>
			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
				<h1>{post().title}</h1>
			</ErrorBoundary>
		</div>
	);
}
\`\`\`
</div>
</TabsCodeBlocks>
```

amirhhashemi replied:
I added a Javascript version to all examples.

I noticed in other sections of the docs, the tabs are labeled "ts" and "js"  instead of "JavaScript" and "TypeScript". I think the full names are more reasonable. I can update the other sections if you'd like.

atilafassina reviewed (commented):
Great work!!

I just had those 2 small comments / nitpicks

amirhhashemi committed:
Add Javascript codes

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

amirhhashemi committed:
update

amirhhashemi committed:
update

amirhhashemi committed:
update title

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide

kodiakhq[bot] committed:
Merge branch 'main' into add-data-fetching-guide
